c_glib/arrow-dataset-glib/file_base.h

```diff
@@ -0,0 +1,170 @@
+struct _GADFileFormatClass
+{
+  GObjectClass parent_class;
+};
+
+gchar *gad_file_format_get_type_name(GADFileFormat *format);
+
+gboolean gad_file_format_is_supported(GADFileFormat *format,
+                                      GADFileSource *source,
+                                      GError **error);
+
+GArrowSchema *gad_file_format_inspect(GADFileFormat *format,
+                                      GADFileSource *source,
+                                      GError **error);
+
+GList *gad_file_format_scan_file(GADFileFormat *format,
+                                 GADFileSource *source,
+                                 GADScanOptions *options,
+                                 GADScanContext *context,
+                                 GError **error);
+
+GADFragment *gad_file_format_make_fragment(GADFileFormat *format,
+                                           GADFileSource *location,
+                                           GADScanOptions *options,
+                                           GError **error);
+
+/* arrow::dataset::FileFragment */
+
+#define GAD_TYPE_FILE_FRAGMENT (gad_file_fragment_get_type())
+G_DECLARE_DERIVABLE_TYPE(GADFileFragment,
+                         gad_file_fragment,
+                         GAD,
+                         FILE_FRAGMENT,
+                         GADFragment)
+struct _GADFileFragmentClass
+{
+  GADFragmentClass parent_class;
+};
+
+GADFileFragment *gad_file_fragment_new(GADFileSource *source,
+                                       GADFileFormat *format,
+                                       GADScanOptions *options);
+const GADFileSource *gad_file_fragment_get_source(GADFileFragment *fragment);
+const GADFileFormat *gad_file_fragment_get_format(GADFileFragment *fragment);
+
+/* arrow::dataset::FileSystemDataset */
+
+#define GAD_TYPE_FILE_SYSTEM_DATASET (gad_file_system_dataset_get_type())
+G_DECLARE_DERIVABLE_TYPE(GADFileSystemDataset,
+                         gad_file_system_dataset,
+                         GAD,
+                         FILE_SYSTEM_DATASET,
+                         GADDataset)
+struct _GADFileSystemDatasetClass
+{
+  GADDatasetClass parent_class;
+};
+
+GADFileSystemDataset *
+gad_file_system_dataset_new(GArrowSchema *schema,
+                            GADExpression *root_partition,
+                            GADFileFormat *format,
+                            GArrowFileSystem *file_system,
+                            GArrowFileInfo *infos);
+
+GADFileSystemDataset *
+gad_file_system_dataset_new_partitions(GArrowSchema *schema,
+                                       GADExpression *root_partition,
+                                       GADFileFormat *format,
+                                       GArrowFileSystem *file_system,
+                                       GArrowFileInfo *infos,
+                                       GADExpression **partitions,
+                                       gsize n_partitions);
+
+/* TODO: GArrowPathForest */
```

Comment:
I'll propose another pull-request to add `GArrowPathForest`.

Reply:
I'll remove `GArrowPathForest` because we decided that we don't use it in #6591.